chunky/src/test/se/llbit/chunky/renderer/scene/ChunkReaderTest.java

```diff
@@ -0,0 +1,186 @@
+import java.util.stream.Stream;
+import org.junit.Test;
+import se.llbit.chunky.chunk.BlockPalette;
+import se.llbit.chunky.chunk.ChunkData;
+import se.llbit.chunky.renderer.scene.ChunkReader.ReadOnlyChunkData;
+import se.llbit.chunky.world.Chunk;
+import se.llbit.chunky.world.ChunkPosition;
+import se.llbit.chunky.world.World;
+import se.llbit.nbt.CompoundTag;
+
+public class ChunkReaderTest {
+
+  @Test
+  public void forEach() {
+    List<ChunkPosition> chunksToLoad = Stream
+        .of(ChunkPosition.get(1, 1), ChunkPosition.get(1, 2), ChunkPosition.get(1, 3))
+        .collect(
+            Collectors.toList());
+    ChunkData chunkData1 = new MockChunkData("A");
+    ChunkData chunkData2 = new MockChunkData("B");
+    Iterator<ChunkData> iterator = Arrays.asList(chunkData1, chunkData2).iterator();
+    MockWorld world = new MockWorld();
+    ListeningExecutorService executorService = MoreExecutors.newDirectExecutorService();
+    ChunkReader chunkReader = new ChunkReader(iterator::next, world,
+        new BlockPalette(new HashMap<>(), new ArrayList<>()), chunksToLoad,
+        executorService);
+
+    List<ChunkPosition> consumedPositions = new ArrayList<>();
+    List<ReadOnlyChunkData> consumedChunkData = new ArrayList<>();
+    List<Integer> consumedProgress = new ArrayList<>();
+    chunkReader.forEach((position, chunkData, progress) -> {
+      consumedPositions.add(position);
+      consumedChunkData.add((ReadOnlyChunkData) chunkData);
+      consumedProgress.add(progress);
+    });
+
+    assertThat(consumedPositions)
+        .containsExactly(ChunkPosition.get(1, 1), ChunkPosition.get(1, 2), ChunkPosition.get(1, 3));
+    assertThat(consumedChunkData.stream().map(roc -> roc.chunkData).collect(Collectors.toList()))
+        .containsExactly(chunkData1, chunkData2, chunkData1).inOrder();
+    assertThat(world.mockChunk.chunksCalledWith)
+        .containsExactly(chunkData1, chunkData2, chunkData1).inOrder();
+    assertThat(consumedProgress).containsExactly(1, 2, 3).inOrder();
+    assertThat(executorService.isShutdown()).isTrue();
+  }
+
+  private static final class MockChunk extends Chunk {
+
+    private final List<ChunkData> chunksCalledWith;
+
+    public MockChunk(World world) {
+      super(ChunkPosition.get(0, 0), world);
+      this.chunksCalledWith = new ArrayList<>();
+    }
+
+    @Override
+    public synchronized ChunkData getChunkData(ChunkData reuseChunkData, BlockPalette palette) {
+      chunksCalledWith.add(reuseChunkData);
+      return reuseChunkData;
+    }
+  }
+
+  private static final class MockWorld extends World {
+
+    private final MockChunk mockChunk;
+
+    protected MockWorld() {
+      super("Mock World", null, 1, Collections.emptySet(), false, 0, 0);
+      mockChunk = new MockChunk(this);
+    }
+
+    @Override
+    public synchronized Chunk getChunk(ChunkPosition pos) {
+      return mockChunk;
+    }
+  }
+
+  private static final class MockChunkData implements ChunkData {
```

Comment:
Instead of micking empty chunk data you could use `EmptyChunkData`. In the future we could maybe even mock some water scenarios (or load _actual_ chunks) and test if finalization works?

Reply:
EmptyChunkData doesn't provide a public constructor and instead provides a single instance. I want to assert that the two chunk data objects are properly being swapped back and forth, so I need different instances.